### ISSUE TITLE
ncat server with self-signed certificate accepts connections from ncat client without cetificate

### ISSUE THREAD

**Issue by JNA4** [nmap/nmap#1898]
When starting a ncat server with self-signed certificate i was still able to connect to the server as a ncat client without using the certificate. The command to start the server:

ncat --listen --ssl --ssl-cert ca-crt.pem --ssl-key ca.key -vv --broker -p 4001

The command for the client:

ncat --ssl -vvv (hostname) 4001

The response i got back had the line:

Ncat: Certificate verification failed (self signed certificate).

But the connection was still created and i was able to send commands to another device that was also connected to the server.

I tried also running the server command without the first --ssl but that didn't work as well.

I couldn't find anywhere any examples for working with a ncat server with a self-signed certificate.

**Comment by nnposter** [comment]
You want to use options `--ssl-verify` and `--ssl-trustfile` on the client side.

**Comment by JNA4** [comment]
It works with --ssl-verify and --ssl-trustfile.
The problem is that it also works when i open the client without them...

**Comment by nnposter** [comment]
You are probably commingling different things. Originally you reported the following issue:
>The response i got back had the line:
>
>Ncat: Certificate verification failed (self signed certificate).
>
>But the connection was still created and i was able to send commands to another device that was also connected to the server.

This is to be expected. By default the client produces the above-noted warning but proceeds with the connection anyway. If you do not want ncat to make this connection unless the server certificate is validated then you should be using the two options I mentioned. The first one is forcing the client to validate the server certificate and the second one is specifying which authorities are considered trusted.

If you are instead trying to achieve something else then please rephrase the issue.

**Comment by JNA4** [comment]
You are right, the 'Certificate verification failed' line is irrelevant here.
Let me rephrase what my issue is:
I am trying to prevent unauthorized clients from connecting to my server.
I opened my server with the line mentioned above. When i tried to connect to it with a client without the certificate it was still able to connect. What do i need to do to make sure that only clients with certificate could connect to my server?

**Comment by nnposter** [comment]
I do not believe that ncat supports certificate authentication of the client. While not ideal, you might be able to put something like `socat` or `haproxy` in front of the broker.

**Comment by Stefan-mcp** [comment]
This seems more a use case for ssh, as you're seemingly looking for a mutually authenticated, transport secured, machine to machine connection.
I think ncats ssl support is more aimed at transport security and compatibility rather than providing a full auth/transport secured remote shell, ssh's got that covered already so it wouldnt be a good use of dev effort

**Comment by dmiller-nmap** [comment]
Client SSL certificate authentication is not planned for Ncat.
